content/blog/2017/2017-01-18-refactoring-conditional-to-pipeline.adoc

```diff
@@ -0,0 +1,475 @@
+
+.Pipeline
+[pipeline]
+----
+// Declarative //
+when {
+    // A or B
+    expression { return A || B }
+}
+steps {
+    /* step */
+}
+
+// Script //
+// A or B
+if (A || B) {
+    /* step */
+}
+----
+
+=== Tokens
+
+Tokens can be considerably more work than conditions.
+There are more of them and they cover a much broader range of behaviors.
+The previous example showed one of the simpler cases, accessing a build parameter,
+where the token has a direct equivalent in Pipeline.
+However, many tokens don't have direct equivalents,
+some take a parameters (adding to their complexity),
+and some provide information that is simply not exposed in pipeline yet.
+So, determining how to migrate tokens needs to be done on case-by-case basis.
+
+Let's look at a few examples.
+
+==== "FILE" token
+
+*Expands to the contents of a file. The file path is relative to the build workspace root.*
+
+_${FILE,path="PATH"}_
+
+This token maps directly to the `readFile` step.
+The only difference is the file path for `readFile` is relative to the
+current working directory on the agent, but that is the workspace root by default.
+No problem.
+
+.Pipeline
+[pipeline]
+----
+// Declarative //
+when {
+    expression { return readFile('pom.xml').contains('mycomponent') }
+}
+steps {
+    /* step */
+}
+
+// Script //
+if (readFile('pom.xml').contains('mycomponent')) {
+    /* step */
+}
+----
+
+
+==== GIT_BRANCH
+
+*Expands to the name of the branch that was built.*
+
+*Parameters* (descriptions omitted): _all_, _fullName_.
+
+This information may or may not be exposed in Pipeline.  If you're using
+multibranch pipeline plugin `env.BRANCH_NAME` give similar basic information, but doesn't
+offer the parameters.  There are also
+link:https://issues.jenkins-ci.org/browse/JENKINS-35230[several]
+link:https://issues.jenkins-ci.org/browse/JENKINS-24141[issues]
+filed around `GIT_*` tokens in Pipeline.
+Until they are addressed fully, we can follow the pattern shown in
+link:https://github.com/jenkinsci/pipeline-examples/blob/master/pipeline-examples/gitcommit/gitcommit.groovy[pipeline-examples],
+executing a shell to get the information we need.
+
+
+.Pipeline
+[source, groovy]
+----
+GIT_BRANCH = sh(returnStdout: true, script: 'git rev-parse --abbrev-ref HEAD').trim()
+----
+
+==== CHANGES_SINCE_LAST_SUCCESS
+
+*Displays the changes since the last successful build.*
+
+*Parameters* (descriptions omitted):
+_reverse_, _format_, _changesFormat_, _showPaths_, _pathFormat_,
+_showDependencies_, _dateFormat_, _regex_, _replace_, _default_.
+
+Not only is the information provided by this token not exposed in Pipeline,
+the token has ten optional parameters, including format strings and regular expression
+searches. There are some examples,
+link:http://stackoverflow.com/questions/38084806/how-to-get-the-changes-since-the-last-successful-build-in-jenkins-pipeline[this answer on stackoverflow.com],
```

Comment:
Yeah, I agree that the code in that SO answer is ugly, and very un-Groovy, but it probably works fine — I've found that using `rawBuild` on rare occasions to obtain changelogs, or test result info, is a quick and pragmatic way of implementing things like this.

Using the CLI would require hardcoding credentials into your `Jenkinsfile` (or ensuring that an SSH private key is available on the right agent), downloading the CLI JAR, figuring out what the `RANGE` argument should be between the current and the last-successful build (if that's even possible), running the CLI, and then parsing the changelog results from XML into your desired format (the plain text output is overly verbose).

Furthermore, the documentation for the `list-changes` CLI command is wrong, and it doesn't even work with Pipeline jobs — it's restricted to finding `AbstractBuild`s rather than `Run`s… :disappointed:

(off to JIRA I go)